File: kotlin/src/main/kotlin/20.ValidParentheses.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*

// https://leetcode.com/problems/valid-parentheses/

val TO_LEFT = mapOf(
    '}' to '{',
    ']' to '[',
    ')' to '('
)

class ValidParentheses {
  fun isValid(s: String): Boolean {
    if(s.length % 2 != 0) return false

    val lefts = Stack<Char>()
    for(ch in s)
      if(TO_LEFT.containsKey(ch)){
        if(lefts.empty() || lefts.pop() != TO_LEFT[ch])
          return false
      } else
        lefts.push(ch)

    return lefts.isEmpty()
  }


  @Test
  fun test1(){
    assertEquals(true, isValid("()"))
  }

  @Test
  fun test2(){
    assertEquals(true, isValid("()[]{}"))
  }

  @Test
  fun test3(){
    assertEquals(false, isValid("([)]"))
  }

  @Test
  fun test4(){
    assertEquals(true, isValid("{[()]}[]"))
  }
}

File: python/20_ValidParentheses.py
from unittest import TestCase

# https://leetcode.com/problems/valid-parentheses/submissions/


class ValidParentheses(TestCase):
    TO_LEFT = {
        ')': '(',
        ']': '[',
        '}': '{',
    }

    def isValid(self, s):
        lefts = []
        for ch in s:
            if self.TO_LEFT.get(ch):
                if not lefts or lefts.pop() != self.TO_LEFT[ch]:
                    return False
            else:
                lefts.append(ch)
        return len(lefts) == 0

    def test1(self):
        self.assertEqual(True, self.isValid("()[]{}"))

    def test2(self):
        self.assertEqual(True, self.isValid("{[()]}[]"))

    def test3(self):
        self.assertEqual(False, self.isValid("[(])"))


